_posts/2025-07-15-fraud-detection-e2e.md

```diff
@@ -0,0 +1,597 @@
+
+```python
+model = build_model(feature_indexes)
+model.fit(x_train, y_train, epochs=2, validation_data=(x_val, y_val), class_weight=class_weights)
+save_model(x_train, model, model_path)  # Exports to ONNX
+```
+
+By structuring the training step this way, the pipeline ensures that the model is trained on the same features that will be available at inference time, supporting a robust and reproducible MLOps workflow.
+
+### 4. Model Registration
+
+Once the model is trained, it's important to track, version, and manage it before deploying to production. This is where the [Kubeflow Model Registry](https://www.kubeflow.org/docs/components/model-registry/) comes in. The Model Registry acts as a centralized service for managing machine learning models and their metadata, making it easier to manage deployments, rollbacks, and audits.
+
+#### Container Image for Model Registration
+
+This pipeline step uses a custom container image built from `pipeline/Containerfile`. The image includes:
+
+* **Kubeflow Pipelines SDK:** For pipeline orchestration and component definitions
+* **Model Registry client:** Python libraries for interacting with the Kubeflow Model Registry
+* **Pipeline orchestration code:** The core pipeline definition and component functions
+
+The container is used as the base image for the `register_model` component, which executes the model registration logic inline within the pipeline definition. This approach allows the registration step to run lightweight operations without requiring a separate, specialized container image.
+
+In the pipeline, the `register_model` component takes the trained model artifact and registers it in the Model Registry. This process includes:
+
+* **Assigning a unique name and version:** The model is registered with a name (e.g., `"fraud-detection"`) and a version, which is typically tied to the pipeline run ID for traceability.
+* **Storing metadata:** Along with the model artifact, metadata such as the model format, storage location, and additional tags or descriptions can be stored for governance and reproducibility.
+* **Making the model discoverable:** Registered models can be easily found and referenced for deployment, monitoring, or rollback.
+
+Here's how the registration step is implemented in the pipeline:
+
+```python
+@dsl.component(base_image=PIPELINE_IMAGE)
+def register_model(model: Input[Model]) -> NamedTuple('outputs', model_name=str, model_version=str):
+    from model_registry import ModelRegistry
+
+    registry = ModelRegistry(
+        server_address="http://model-registry-service.kubeflow.svc.cluster.local",
+        port=8080,
+        author="fraud-detection-e2e-pipeline",
+        user_token="non-used",
+        is_secure=False
+    )
+
+    model_name = "fraud-detection"
+    model_version = "{{workflow.uid}}"
+
+    registry.register_model(
+        name=model_name,
+        uri=model.uri,
+        version=model_version,
+        model_format_name="onnx"
+    )
```

Comment:
here you can actually consider leveraging the `modelSource*` fields from here:

https://github.com/kubeflow/model-registry/blob/367a87fb56e82ce26fb970e0fdf5d159cd4934d8/clients/python/src/model_registry/_client.py#L285-L289

apologies I didn't notice this earlier in your demo code.

@mprahl can recommend field valorization for KFP.

Reply:
Done.